modules/percolator/src/main/java/org/elasticsearch/percolator/QueryAnalyzer.java

```diff
@@ -354,43 +411,133 @@ static Result handleDisjunction(List<Query> disjunctions, int minimumShouldMatch
         return new Result(verified, terms);
     }
 
-    static Set<Term> selectTermListWithTheLongestShortestTerm(Set<Term> terms1, Set<Term> terms2) {
+    static Set<QueryTerm> selectTermListWithTheLongestShortestTerm(Set<QueryTerm> terms1, Set<QueryTerm> terms2) {
+        assert terms1 != null || terms2 != null;
         if (terms1 == null) {
             return terms2;
         } else if (terms2 == null) {
             return terms1;
         } else {
-            int terms1ShortestTerm = minTermLength(terms1);
-            int terms2ShortestTerm = minTermLength(terms2);
-            // keep the clause with longest terms, this likely to be rarest.
-            if (terms1ShortestTerm >= terms2ShortestTerm) {
-                return terms1;
+            // Prefer term based extractions over range based extractions:
+            boolean onlyRangeBasedExtractions = false;
+            for (QueryTerm queryTerm : terms1) {
+                if (queryTerm.range != null) {
+                    onlyRangeBasedExtractions = true;
+                    break;
+                }
+            }
+            for (QueryTerm queryTerm : terms2) {
+                if (queryTerm.range != null) {
+                    onlyRangeBasedExtractions = true;
```

Comment:
it look backwards: this will make `onlyRangeBasedExtractions = true` if any of the extractions returned a range, while the name of the variable suggests that you want to set it to true if all extractions return a range?